docs/02-app/01-building-your-application/09-authentication/index.mdx

```diff
@@ -0,0 +1,533 @@
+4. **Handling Unauthorized Access:**
+   - Redirect unauthorized users to a login or error page as appropriate.
+
+Example middleware file:
+
+```ts filename="middleware.ts" switcher
+import { NextResponse } from 'next/server'
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  const currentUser = request.cookies.get('currentUser')?.value
+
+  if (currentUser) {
+    return NextResponse.redirect(new URL('/dashboard', request.url))
+  }
+  return NextResponse.redirect(new URL('/login', request.url))
+}
+
+export const config = {
+  matcher: ['/((?!api|_next/static|_next/image|.*\\.png$).*)'],
+}
+```
+
+```js filename="middleware.js" switcher
+import { NextResponse } from 'next/server'
+
+export function middleware(request) {
+  const currentUser = request.cookies.get('currentUser')?.value
+
+  if (currentUser) {
+    return NextResponse.redirect(new URL('/dashboard', request.url))
+  }
+  return NextResponse.redirect(new URL('/login', request.url))
+}
+
+export const config = {
+  matcher: ['/((?!api|_next/static|_next/image|.*\\.png$).*)'],
+}
+```
+
+This example uses [`NextResponse.redirect`](https://nextjs.org/docs/app/api-reference/functions/next-response#redirect) for efficient, centralized access control. It handles redirects early in the request pipeline, making it efficient and ideal for broad strategies.
+
+For specific redirection needs, the `redirect` function can be used in Server Components, Route Handlers, and Server Actions to provide more control. This is useful for role-based navigation or context-sensitive scenarios.
+
+```ts filename="/app/page.tsx" switcher
+import { redirect } from 'next/navigation'
+
+export default function Page() {
+  // Logic to determine if a redirect is needed
+  const accessDenied = true
+  if (accessDenied) {
+    redirect('/login')
+  }
+
+  // Define other routes and logic
+}
+```
+
+```js filename="/app/page.jsx" switcher
+import { redirect } from 'next/navigation'
+
+export default function Page() {
+  // Logic to determine if a redirect is needed
+  const accessDenied = true
+  if (accessDenied) {
+    redirect('/login')
+  }
+
+  // Define other routes and logic
+}
+```
+
+This flexibility is crucial for role-specific redirections post-authentication. For example, an admin user can be redirected to an admin dashboard, while a regular user might be sent to a different page. It's also useful for conditional navigation, like redirecting users to complete their profile if necessary.
+
+When setting up authorization, it's important to make sure the main security checks happen where your app accesses or changes data. Middleware can help by doing a first check, but it shouldn't be the only thing keeping your data safe. Always check if a user has the right access when you're actually getting or changing the data. This way, even if someone gets past the first check, your data is still secure at the most important step - when it's being accessed.
```

Comment:
I added this section based on the conve we had with @leerob regarding https://vercel.slack.com/archives/C03KAR5DCKC/p1704943720619029?thread_ts=1704942838.625119&cid=C03KAR5DCKC. What do you think? I kept it brief, is it worth going into more detail?